engine/consensus/matching/engine.go

```diff
@@ -587,32 +600,45 @@ func (e *Engine) sealableResults() ([]*flow.IncorporatedResult, error) {
 
 // matchChunk checks that the number of ResultApprovals collected by a chunk
 // exceeds the required threshold.
-func (e *Engine) matchChunk(resultID flow.Identifier, chunk *flow.Chunk, assignment *chunks.Assignment) bool {
+func (e *Engine) matchChunk(resultID flow.Identifier, chunk *flow.Chunk, assignment *chunks.Assignment) (bool, error) {
 
 	// get all the chunk approvals from mempool
 	approvals := e.approvals.ByChunk(resultID, chunk.Index)
 
-	// only keep approvals from assigned verifiers
+	// only keep approvals from assigned verifiers and if spocks match
 	var validApprovers flow.IdentifierList
-	for approverID := range approvals {
+	for approverID, approval := range approvals {
+		// check if chunk is assigned
 		ok := chmodule.IsValidVerifer(assignment, chunk, approverID)
-		if ok {
-			validApprovers = append(validApprovers, approverID)
+		if !ok {
+			continue
+		}
+
+		// check if spocks match
+		valid, err := e.spockVerifier.VerifyApproval(approval)
+		if err != nil {
+			return false, fmt.Errorf("could get validate spocks: %w", err)
+		}
+		if !valid {
+			e.log.Error().Msg("spock validation failed: spocks do not match")
+			continue
 		}
+
+		// add approver to valid approvers list
+		validApprovers = append(validApprovers, approverID)
```

Comment:
We have to know which receipt the approval is approving for, and group approvals that approving the receipts that has the matching sporks. Otherwise, it's possible the approvals are approving mismatching sporks.

Reply:
Good catch Leo

Reply:
Another point: how can we make sure `validApprovers` for each receipt are all distinct approvers? 
I can see we are going through approvals resulting from `approvals := e.approvals.ByChunk(resultID, chunk.Index)` but are we sure all these no VN can send 2 different approvals with the same resultID and chunk index?

Reply:
`e.approvals` ensures that you can only insert one approval per verifier per chunk